File: idl706linux.x86/idl706/external/callable/calltest.c
/*
 * calltest - Simple demonstration program for Callable IDL.
 *
 *	$Id: //depot/idl/IDL_70/idldir/external/callable/calltest.c#3 $
 *
 * This program uses callable IDL to do the following things:
 *
 *	- Import a 10 element array (equivalent to a FINDGEN(10))
 *        created by our C program into IDL, under the name TMP.
 *      - Have callable IDL total the imported TMP array, storing the
 *        result in an IDL variable named TMP_TOTAL.
 *      - Use callable IDL to print the values of TMP, TMP_TOTAL,
 *        and to produce a plot of TMP.
 *      - Demonstrate that when callable IDL releases the imported
 *        array TMP, that the user registered callback is called.
 *      - Access an IDL variable (TMP_TOTAL) from our C program.
 *	- Create a blocking IDL WIDGET button with the title "Done".
 *	  When the user presses this button, callable IDL is shut down
 *	  and the program exits.
 *
 * Notes:
 *    - For Microsoft Windows, there is a significant amount of
 *      code required to create a window to display IDL's output
 *      window and to handle the additional initialization steps
 *      required on that platform. That code is found in calltest_win.c.
 *    - The PRINTF_LINE macro is defined in calltest.h. It allows this
 *      rogram to generate printf()-like output on both platforms without
 *	using #ifdef lines everywhere.
 */

/*
 *  Copyright (c) 1992-2008, ITT Visual Information Solutions. All
 *  rights reserved. Reproduction by any means whatsoever is prohibited 
 *  without express written permission.
 */

#ifdef WIN32
#include <windows.h>
#endif
#include <stdio.h>
#include "idl_export.h"
#include "calltest.h"







static void free_callback(UCHAR *addr)
/*
 * This function is called by IDL when the imported TMP variable is
 * set to 0, releasing the imported memory.
 */
{
  PRINTF_LINE("IDL released(%p)", addr);
}



static void demo_callable_idl(void)
{
  float f[10];
  int i;
  IDL_VPTR v;
  IDL_MEMINT dim[IDL_MAX_ARRAY_DIM];
  static char *cmds[] = { "tmp_total = total(tmp)",
			  "print,'IDL total is ', tmp_total",
			  "plot,tmp" };
  static char *cmds2[] = { "a = widget_base()",
     "b = widget_button(a, value='Press When Done', xsize=300, ysize=200)",
     "widget_control,/realize, a",
     "dummy = widget_event(a)",
     "widget_control,/destroy, a" };
  

  for (i=0; i < 10; i++) f[i] = (float) i;
  dim[0] = 10;
  PRINTF_LINE("ARRAY ADDRESS(%u)", f);
  if (v = IDL_ImportNamedArray("TMP", 1, dim, IDL_TYP_FLOAT, (UCHAR *) f,
			       free_callback, (void *) 0)) {
    (void) IDL_ExecuteStr("print, tmp");
    (void) IDL_Execute(sizeof(cmds)/sizeof(char *), cmds);
    (void) IDL_ExecuteStr("print, 'This should free the user memory");
    (void) IDL_ExecuteStr("tmp = 0");
    if (v = IDL_FindNamedVariable("tmp_total", FALSE))
      PRINTF_LINE("Program total is %f", v->value.f);
    (void) IDL_Execute(sizeof(cmds2)/sizeof(char *), cmds2);
  }
}







#ifdef WIN32
int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hInstancePrev,
	LPSTR lpszCmndline, int nCmdShow)
/*
 * Windows applications start at WinMain() instead of main().
 */
{
  /* Window's specific initialization */
  if (!CalltestWinMain(hInstance, hInstancePrev, lpszCmndline, nCmdShow))
    return FALSE;

  demo_callable_idl();		/* Use Callable IDL */

  IDL_Cleanup(FALSE);		/* Don't return */
    
  return TRUE;
}
#endif







#ifndef WIN32			/* !WIN32 == Unix */
int main(int argc, char **argv)
{
  IDL_INIT_DATA init_data;

  init_data.options = IDL_INIT_NOCMDLINE;
  if (argc) {
    init_data.options |= IDL_INIT_CLARGS;
    init_data.clargs.argc = argc;
    init_data.clargs.argv = argv;
  }
  if (IDL_Initialize(&init_data)) {
    demo_callable_idl();
    IDL_Cleanup(FALSE);	/* Don't return */
  }

  return 1;			/* Will only get here if IDL_Init() fails. */
}
#endif				/* Unix */

File: idl706linux.x86/idl706/external/callable/calltest.f
C-----------------------------------------------------------------
C  	Routine to print a floating point value from an IDL variable.

	SUBROUTINE PRINT_FLOAT(VPTR)

C       Declare a Fortran Record type that has a compatible form with 
C       the IDL C struct IDL_VARIABLE for a floating point value.
C       Note this structure contains a union which is the size of 
C       the largest data type.  This structure has been padded to
C       support the union.   Fortran records are not part of
C       F77, but most compilers have this option.

	STRUCTURE /IDL_VARIABLE/
           CHARACTER*1 TYPE 
           CHARACTER*1 FLAGS 
           INTEGER*4 PAD	!Pad for largest data type 
           REAL*4 VALUE_F
	END STRUCTURE 

	RECORD /IDL_VARIABLE/ VPTR

	WRITE(*, 10) VPTR.VALUE_F
  10	FORMAT('Program total is: ', F6.2)

	RETURN

	END

C-----------------------------------------------------------------
C  This function will be called when IDL is finished with the 
C  array F.  

       SUBROUTINE FREE_CALLBACK(ADDR)

          INTEGER*4 ADDR

          WRITE(*,20) LOC(ADDR)
  20	  FORMAT ('IDL Released:', I12)

          RETURN

       END

C-----------------------------------------------------------------
C   This program demonstrates how to import data from a Fortran
C   program into IDL, execute IDL statements and obtain data
C   from IDL variables.  

      PROGRAM CALLTEST

C   Some Fortran compilers require external definitions for IDL routines
        EXTERNAL IDL_Init !$pragma C(IDL_Init)    
        EXTERNAL IDL_Cleanup !$pragma C(IDL_Cleanup)    
        EXTERNAL IDL_Execute !$pragma C(IDL_Execute)
        EXTERNAL IDL_ExecuteStr !$pragma C(IDL_ExecuteStr)
        EXTERNAL IDL_ImportNamedArray !$pragma C(IDL_ImportNamedArray)
        EXTERNAL IDL_FindNamedVariable !$pragma C( IDL_FindNamedVariable )


C   Define arguments for IDL_Init routine
        INTEGER*4 ARGC
        INTEGER*4 ARGV(1)
        DATA ARGC, ARGV(1) /2 * 0/
  
C   Define IDL Definitions  for IDL_ImportNamedArray

        PARAMETER (IDL_MAX_ARRAY_DIM = 8)
        PARAMETER (IDL_TYP_FLOAT = 4)

        REAL*4 F(10)
        INTEGER*4 DIM(IDL_MAX_ARRAY_DIM)
        DATA DIM /10, 7*0/
        INTEGER*4 VAR_PTR 	!Address of IDL variable
        EXTERNAL FREE_CALLBACK	!Declare external routine for use as arg

        PARAMETER (MAXLEN=80)   !Maximum character string length
	PARAMETER (N_ELTS=10)	!Number of elements in array F

C  Define commands to be executed by IDL

        CHARACTER*(MAXLEN) CMDS(3)
        DATA CMDS /"tmp2 = total(tmp)",
     &            "print, 'IDL total is ', tmp2",
     &            "plot, tmp"/
        INTEGER*4 CMD_ARGV(10)

C  Define widget commands to be executed by IDL

        CHARACTER*(MAXLEN) WIDGET_CMDS(5)
        DATA  WIDGET_CMDS /"a = widget_base()",
     &    "b = widget_button(a,val='Press When Done',xs=300,ys=200)",
     &    "widget_control, /realize, a",
     &    "dummy = widget_event(a)", 
     &    "widget_control, /destroy, a"/

        INTEGER*4 ISTAT 

C    Null Terminate command strings and store the address
C    for each command string in CMD_ARGV 

        DO I = 1, 3  
           CMDS(I)(MAXLEN:MAXLEN) = CHAR(0)
           CMD_ARGV(I) = LOC(CMDS(I))
        ENDDO

C   Initialize floating point array, equivalent to IDL FINDGEN(10)

        DO I = 1, N_ELTS
           F(I) = FLOAT(I-1)
        ENDDO

C   Print address of F 

	WRITE(*,30) LOC(F)
   30	FORMAT('ARRAY ADDRESS:', I12)

C   Initialize Callable IDL

        ISTAT = IDL_Init(%VAL(0), ARGC, ARGV(1))

        IF (ISTAT .EQ. 1) THEN 

C   Import the floating point array into IDL as a variable named TMP 

          CALL IDL_ImportNamedArray('TMP'//CHAR(0), %VAL(1), DIM, 
     &	       %VAL(IDL_TYP_FLOAT), F, FREE_CALLBACK, %VAL(0))

C   Have IDL print the value of tmp

          CALL IDL_ExecuteStr('PRINT, TMP'//CHAR(0))

C   Execute a short sequence of IDL statements from a string array 

          CALL IDL_Execute(%VAL(3), CMD_ARGV)

C   Set tmp to zero, causing IDL to release the pointer to the
C   floating point array.

          CALL IDL_ExecuteStr('TMP = 0'//CHAR(0))

C   Obtain the address of the IDL variable containing the
C   the floating point data 

          VAR_PTR = IDL_FindNamedVariable('TMP2'//CHAR(0), %VAL(0)) 

C   Call a Fortran routine to print the value of the IDL tmp2 variable 
          CALL PRINT_FLOAT(%VAL(VAR_PTR))


C    Null Terminate command strings and store the address
C    for each command string in CMD_ARGV 

          DO I = 1, 5  
             WIDGET_CMDS(I)(MAXLEN:MAXLEN) = CHAR(0)
             CMD_ARGV(I) = LOC(WIDGET_CMDS(I))
          ENDDO

C   Execute a small widget program.  Pressing the button allows
C   the program to end 

          CALL IDL_Execute(%VAL(5), CMD_ARGV)

C   Shut down IDL
          CALL IDL_Cleanup(%VAL(0))

        ENDIF

      END
    


